tool/src/org/antlr/v4/codegen/UnicodeEscapes.java

```diff
@@ -6,33 +6,47 @@
 
 package org.antlr.v4.codegen;
 
+import org.antlr.v4.codegen.target.*;
+
 /**
  * Utility class to escape Unicode code points using various
- * languages' syntaxes.
+ * languages' syntax.
  */
-public abstract class UnicodeEscapes {
-	static public void appendJavaStyleEscapedCodePoint(int codePoint, StringBuilder sb) {
-		if (Character.isSupplementaryCodePoint(codePoint)) {
-			// char is not an 'integral' type, so we have to explicitly convert
-			// to int before passing to the %X formatter or else it throws.
-			sb.append(String.format("\\u%04X", (int)Character.highSurrogate(codePoint)));
-			sb.append(String.format("\\u%04X", (int)Character.lowSurrogate(codePoint)));
-		}
-		else {
-			sb.append(String.format("\\u%04X", codePoint));
-		}
+public class UnicodeEscapes {
+	public static String escapeCodePoint(int codePoint, String language) {
+		StringBuilder result = new StringBuilder();
+		appendEscapedCodePoint(result, codePoint, language);
+		return result.toString();
 	}
 
-	static public void appendPythonStyleEscapedCodePoint(int codePoint, StringBuilder sb) {
-		if (Character.isSupplementaryCodePoint(codePoint)) {
-			sb.append(String.format("\\U%08X", codePoint));
+	public static void appendEscapedCodePoint(StringBuilder sb, int codePoint, String language) {
+		switch (language) {
+			case CSharpTarget.key:
```

Comment:
actually, hang on a second.  why isn't this just

```
switch ( language ) {
    case "Cpp": ...
    case "Java": ...
...
}
```

We have a requirement that XTarget be language X.  That way we can simplify creating the  constants by looking at the class name in the target constructor.  In other words, `Python3Target.key` must always be `Python3`.

Reply:
To prevent future potential mistakes in string literals if these constants will be used something else. For instance, `BaseRuntimeTest` and other similar classes can use these constants as well.

Reply:
in general you're following a great principle here which is to use the type system versus strings, but code generators are specifically about flexibility and besides. People are always able to make errors and strings even here:

```
	public final static String key = "CSharp";
```

As long as we are going for flexibility, we might as well remove all of the extra code you've added to each target and simplify the switch statement. 

literally the only place we use the string is in the switch statement, if intellij is giving me the straight story. I can delete many lines of code from your PR and simplify multiple expressions by going back to a string.